tee_supplicant/src/tee_supplicant.c

```diff
@@ -135,6 +140,451 @@ static int load_ta(uint32_t num_params, struct tee_param *params)
+{
+	int rc, fd;
+	off_t len;
+	void *ptr = NULL;
+	struct fs_file_t *file;
+
+	if (num_params != 1  || (params[0].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_VALUE_INPUT)
+	{
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	fd = params[0].b;
+	len = params[0].c;
+	ptr = z_get_fd_obj(fd, NULL, 0);
+
+	if (!ptr) {
+		LOG_ERR("descriptor %d not found", fd);
+		return TEEC_ERROR_ITEM_NOT_FOUND;
+	}
+
+	/*HACK: this ptr is of hidden posix_fs_desc type, begins w/ file attr */
+	file = (struct fs_file_t *)ptr;
+	rc = fs_truncate(file, len);
+	if (rc < 0) {
+		LOG_ERR("failed to truncate (%d)", rc);
+		return TEEC_ERROR_GENERIC;
+	}
+
+	return TEEC_SUCCESS;
+}
+
+static int tee_fs_remove(size_t num_params, struct tee_param *params)
+{
+	struct tee_shm *shm;
+	char *name, path[128 + 4] = "/tee";
+
+	if (num_params != 2) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	if ((params[0].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_VALUE_INPUT ||
+	    (params[1].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_MEMREF_INPUT)
+	{
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	/*TODO do all safety checks */
+	shm = (struct tee_shm *)params[1].c;
+
+	if (!shm || !shm->addr) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	name = shm->addr;
+	strncat(path, name, 128);
+
+	if (unlink(path) < 0) {
+		if (errno == ENOENT) {
+			return TEEC_ERROR_ITEM_NOT_FOUND;
+		}
+		LOG_ERR("failed to unlink %s (%d)", path, -errno);
+		return TEEC_ERROR_GENERIC;
+	}
+
+	/*TODO: cleanup empty directories */
+	return TEEC_SUCCESS;
+}
+
+static int tee_fs_rename(size_t num_params, struct tee_param *params)
+{
+	char *name, path[128 + 4] = "/tee";
+	char *new_name, new_path[128 + 4] = "/tee";
+	struct tee_shm *shm;
+
+	if (num_params != 3) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	if ((params[0].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_VALUE_INPUT ||
+	    (params[1].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_MEMREF_INPUT ||
+	    (params[2].attr & TEE_PARAM_ATTR_TYPE_MASK) !=
+		TEE_PARAM_ATTR_TYPE_MEMREF_INPUT)
+	{
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	/*TODO do all safety checks */
+	shm = (struct tee_shm *)params[1].c;
+
+	if (!shm || !shm->addr) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	name = shm->addr;
+	strncat(path, name, 128);
+
+	shm = (struct tee_shm *)params[2].c;
+
+	if (!shm || !shm->addr) {
+		return TEEC_ERROR_BAD_PARAMETERS;
+	}
+
+	new_name = shm->addr;
+	strncat(new_path, new_name, 128);
+
+	if (!params[0].b) {
+		struct stat buf;
+		if (stat(new_path, &buf) < 0) {
+			return TEEC_ERROR_ACCESS_CONFLICT;
+		}
+	}
+
+	if (rename(path, new_path) < 0) {
+		LOG_ERR("failed to rename %s -> %s (%d)",
+			path, new_path, -errno);
+		return TEEC_ERROR_GENERIC;
+	}
+
+	return TEEC_SUCCESS;
+}
+
+static int tee_fs_opendir(size_t num_params, struct tee_param *params)
+{
+	struct tee_shm *shm;
+	char *name, path[128 + 4] = "/tee";
```

Comment:
Ditto

Reply:
fixed in [fbb0dd0](https://github.com/dsemenets/zephyr-optee-client/pull/1/commits/fbb0dd06f8161252c3a15bd52612e52c1490364a)